ap_cis = readRDS("data/out/combined/Ap/geno_cis_comp.RDS")
ap_cis$prevGenoCis = ap_cis$prevGenoCis %>% mutate(p_adj=p.adjust(p,method = "fdr"))
ap_cis$hmmGenoCis = ap_cis$hmmGenoCis %>% mutate(p_adj=p.adjust(p,method = "fdr"))
ap_combined = ap_cis$prevGenoCis %>% inner_join(ap_cis$hmmGenoCis,by=c("gene"),suffix=c(".old",".hmm"))
ap_combined = ap_combined %>% mutate(sig_both = beta.old * beta.hmm > 0 & p_adj.old < 0.05 & p_adj.hmm < 0.05)
ap_combined_1000 = ap_combined %>% full_join(cis_test_1000,by=c("gene"="transcript"))
#ap_combined %>% ggplot(aes(y=beta.old,x=))
one_pot_bulk =cross_data$A$cis$cis_test_with_disp_expr %>% full_join(cis_test_1000,by=c("transcript"="transcript"))
sum(one_pot_bulk$FDR < 0.05 & one_pot_bulk$Beta * one_pot_bulk$coefficient > 0,na.rm=T)
nrow(ap_combined %>% filter(p_adj.old < 0.05))
nrow(ap_combined %>% filter(p_adj.hmm < 0.05))
table(ap_combined$sig_both)
ap_combined = ap_combined %>% inner_join(genes_name_trans,by=c("gene"="gene_id"))
ap_combined =ap_combined %>% dplyr::select(gene,gene_name,beta.old,p.old,p_adj.old,beta.hmm,p.hmm,p_adj.hmm,sig_both, gene_name)
ap_combined$sig_both = as.numeric(ap_combined$sig_both)
ap_combined2 = ap_combined 
colnames(ap_combined2) = c("transcript","gene","beta (whole-genome)","p-value (whole-genome)","adjusted p-value (whole-genome)",
                          "beta (HMM)","p-value (HMM)","adjusted p-value (HMM)","significant using both genotyping methods")
#ap_combined %>% write_tsv("tables/s6.tsv")
openxlsx::write.xlsx(ap_combined2,"tables/s6.xlsx")
